test/conftest.py

```diff
@@ -796,54 +794,56 @@ async def QueueLen(self, stream):
 
     ### Sandbox
 
-    # Add shell commands to submit to the sandbox after it's created.
-    def add_shell_cmds(self, args: list[bytes]):
-        self.pending_shell_cmds += args
-
-    def is_shell_cmds(self, cmds):
-        shell_list = ["bash", "sh"]
-        if len(cmds) == 1 and cmds[0] in shell_list:
-            return True
-        else:
-            return False
-
     async def SandboxCreate(self, stream):
         request: api_pb2.SandboxCreateRequest = await stream.recv_message()
-        # Not using asyncio.subprocess here for Python 3.7 compatibility.
-        self.sandbox = subprocess.Popen(
-            request.definition.entrypoint_args,
+        if self.is_shell_cmds(request.definition.entrypoint_args):
+            self.is_shell = True
+        self.sandbox = await asyncio.subprocess.create_subprocess_exec(
+            *request.definition.entrypoint_args,
             stdout=asyncio.subprocess.PIPE,
             stderr=asyncio.subprocess.PIPE,
             stdin=asyncio.subprocess.PIPE,
         )
+
         self.sandbox_defs.append(request.definition)
-        if self.is_shell_cmds(request.definition.entrypoint_args) and len(self.pending_shell_cmds) == 0:
-            raise Exception(
-                f"The sandbox commands is ${request.definition.entrypoint_args} but there is no pending command to exit the shell. Call 'pending_shell_cmds' to add pending commands."
-            )
-        for cmd in self.pending_shell_cmds:
-            self.sandbox.stdin.write(cmd)
-        self.sandbox.stdin.flush()
+
         await stream.send_message(api_pb2.SandboxCreateResponse(sandbox_id="sb-123"))
 
+    def is_shell_cmds(self, cmds):
+        shell_list = ["bash", "sh"]
+        if len(cmds) == 1 and cmds[0] in shell_list:
+            return True
+        else:
+            return False
+
     async def SandboxGetLogs(self, stream):
         request: api_pb2.SandboxGetLogsRequest = await stream.recv_message()
+        io: asyncio.StreamReader
+        if self.is_shell:
+            # sends an empty message to simulate PTY
+            await stream.send_message(
+                api_pb2.TaskLogsBatch(items=[api_pb2.TaskLogs(data="", file_descriptor=request.file_descriptor)])
+            )
+
         if request.file_descriptor == api_pb2.FILE_DESCRIPTOR_STDOUT:
             # Blocking read until EOF is returned.
-            data = self.sandbox.stdout.read()
+            io = self.sandbox.stdout
```

Comment:
nit: will shadow the `io` stdlib library if it's imported. Maybe just `file`, or `f`?

Reply:
just had a boomer moment where i was about to "well, actually" you on `file` also being a builtin name, but it turns out they changed that in Python 3 🙈

Reply:
haha nice. I've written <1000 lines of Python 2 and never really had to mentally transition. TIL https://portingguide.readthedocs.io/en/latest/builtins.html#removed-file